ffi/diplomat/src/calendar.rs

```diff
@@ -32,4 +37,56 @@ pub mod ffi {
                 .into()
         }
     }
+
+    #[diplomat::opaque]
+    #[diplomat::rust_link(icu::calendar::AnyCalendar, Enum)]
+    pub struct ICU4XCalendar(pub Arc<AnyCalendar>);
+
+    impl ICU4XCalendar {
+        /// Creates a new [`ICU4XGregorianDateTime`] from the specified date and time.
+        #[diplomat::rust_link(icu::calendar::AnyCalendar::try_new_unstable, FnInStruct)]
+        pub fn try_new(
+            provider: &ICU4XDataProvider,
+            locale: &ICU4XLocale,
+        ) -> DiplomatResult<Box<ICU4XCalendar>, ICU4XError> {
+            use icu_provider::serde::AsDeserializingBufferProvider;
+            let provider = provider.0.as_deserializing();
+
+            let kind = match (&locale.0).try_into() {
+                Ok(k) => k,
+                Err(e) => return Err(ICU4XError::from(e)).into(),
+            };
+
+            AnyCalendar::try_new_unstable(&provider, kind)
+                .map(|df| Box::new(ICU4XCalendar(Arc::new(df))))
+                .map_err(Into::into)
+                .into()
+        }
+    }
+
+    #[diplomat::opaque]
+    /// An ICU4X DateTime object capable of containing a Gregorian date and time.
+    #[diplomat::rust_link(icu::calendar::DateTime, Struct)]
+    pub struct ICU4XDateTime(pub DateTime<Arc<AnyCalendar>>);
+
+    impl ICU4XDateTime {
+        /// Creates a new [`ICU4XDateTime`] representing the ISO date and time
+        /// given but in a given calendar
+        #[diplomat::rust_link(icu::calendar::DateTime::new_iso_datetime, FnInStruct)]
+        pub fn try_new_from_iso_in_calendar(
+            year: i32,
+            month: u8,
+            day: u8,
+            hour: u8,
+            minute: u8,
+            second: u8,
```

Comment:
Thought: I've said this before, and I'll say it again; I think we should allow for fractional seconds. We support fractional seconds in code and patterns, and there are real use cases for them, and we shouldn't demote them to some future second-class constructor. I acknowledge that we didn't end up doing that in the Rust API.

Reply:
To be clear, the intent is for them to be mutated in, not in a constructor. I can add a mutation API, I was trying to add the minimal APIs for the demo.

I think they would be fine in a constructor that took in a struct, but otherwise with just ordered parameters it's unclear what the last parameter is from context (and it makes the identity of the other parameters confusing), which is why I've been reluctant to have this in the ctor without having `with_fractional_seconds` in its name

Reply:
Also I think in our final version we should also have a from_codes ctor, I just didn't add it here because I was going for minimal necessary for the demo

Reply:
I guess the other option would be to make the `second` argument an `f64`... but then we get into ryu stuff pretty quickly (although we could likely use the fast path to avoid ryu).